src/v/storage/segment_set.cc

```diff
@@ -197,6 +199,83 @@ is_last_segment(segment* s, std::optional<ss::sstring> last_clean_segment) {
                 == std::string(last_clean_segment.value());
 }
 
+ss::future<std::optional<segment_set>>
+maybe_create_contiguous_segment_set(segment_set::underlying_t segs) {
+    if (segs.empty()) {
+        co_return std::nullopt;
+    }
+    segment_set seg_set = segment_set(std::move(segs));
+    auto min_offset = seg_set.front()->offsets().get_base_offset();
+    // Max offset may not necessarily be seg_set.back()'s dirty_offset, due to
+    // descending ordering w/r/t dirty offset in `segment_ordering`.
+    auto max_offset = (*std::ranges::max_element(
+                         seg_set,
+                         std::less<>{},
+                         [](const auto& s) {
+                             return s->offsets().get_dirty_offset();
+                         }))
+                        ->offsets()
+                        .get_dirty_offset();
+    segment_set::underlying_t ignored_segs;
+    for (auto it = seg_set.begin(); it != seg_set.end();) {
+        if (it != seg_set.begin()) {
+            auto& s = *it;
+            auto& prev = *std::prev(it);
+            if (
+              prev->offsets().get_dirty_offset()
+              >= s->offsets().get_base_offset()) {
+                vlog(
+                  stlog.info,
+                  "Base offset {} of segment {} is < dirty offset {} of "
+                  "previous segment {} after recovery. This is very likely a "
+                  "segment that was not successfully removed during adjacent "
+                  "merge compaction.",
+                  s->offsets().get_base_offset(),
+                  s->filename(),
+                  prev->offsets().get_dirty_offset(),
+                  prev->filename());
+                ignored_segs.push_back(s);
+                it = seg_set.erase(it);
+                continue;
+            }
+            // At this point we have greedily selected the segment `s`- ensure
+            // offset continuity in the same pass
```

Comment:
nit: Looking at this code fresh, this comment about "same pass" seems out of place. Is "same pass" alluding to another implementation that had multiple passes?